### PR TITLE
Fix `unit-no-unknown` false positives for the second and subsequent `image-set()` with `x` descriptor

### PR DESCRIPTION
**Pull request by romainmenke**
> Which issue, if any, is this issue related to?

See : https://github.com/stylelint/stylelint/issues/6848
Supersedes : https://github.com/stylelint/stylelint/pull/6726

> Is there anything in the PR that needs further explanation?

- ~~this will fail because it uses unreleased features of the media query parser~~
- also happens to resolve #6726

before :

```
npm run benchmark-rule unit-no-unknown true
Warnings: 0
Mean: 143.3089323125 ms
Deviation: 10.593647978441565 ms
```

after : 

```
npm run benchmark-rule unit-no-unknown true
Warnings: 0
Mean: 109.86287500000002 ms
Deviation: 9.320724134887374 ms
```

### CODE REVIEW

**Comment by romainmenke** [comment]
@ybiquitous I intended to just change the parser here, but this change also had the side effect of fixing an issue.

I propose changing the title and adding the correct changelog for that:

```md
Fix `unit-no-unknown` false positives for the second and subsequent `image-set()` with `x` descriptor
```

I would also add `chalkygames123` as a co-author as they wrote the tests that I used to verify the fix in https://github.com/stylelint/stylelint/pull/6726

**Comment by ybiquitous** [comment]
Yes, that's a good idea on https://github.com/stylelint/stylelint/pull/6879#issuecomment-1572262355 

Please update them!

File: lib/rules/unit-no-unknown/index.js [hunk 1]
```diff
@@ -48,122 +58,149 @@ const rule = (primary, secondaryOptions) => {
...
+			if (!tokens) return;
 
-			check(atRule, atRule.params, atRuleParamIndex);
+			parseFromTokens(tokens).forEach((mediaQuery) => {
+				const state = {
+					ignored: false,
+					allowX: false,
+				};
+
+				mediaQuery.walk((entry) => {
```
Comment (ybiquitous) [inline lib/rules/unit-no-unknown/index.js:149]: [ask] Can `entry.state` be empty? If no, I guess we could simplify the code by early return like this:

```suggestion
				mediaQuery.walk((entry) => {
					if (!entry.state) return;
```

File: lib/rules/unit-no-unknown/index.js [hunk 2]
```diff
@@ -48,122 +58,149 @@ const rule = (primary, secondaryOptions) => {
...
+
+					if (isMediaFeature(entry.node)) {
+						const name = entry.node.getName();
+
+						if (entry.state && RESOLUTION_MEDIA_FEATURE_NAME.test(name)) {
+							entry.state.allowX = true;
+						}
+					}
+
+					if (isFunctionNode(entry.node) || isTokenNode(entry.node)) {
```
Comment (ybiquitous) [inline lib/rules/unit-no-unknown/index.js:160]: [ask] Does converting to `else if` make sense to reduce extra evaluations?

```suggestion
					} else if (isFunctionNode(entry.node) || isTokenNode(entry.node)) {
```

File: lib/rules/unit-no-unknown/index.js (hunk 1 above)
Comment (romainmenke) [reply lib/rules/unit-no-unknown/index.js:149]: It should never be empty unless there is a bug or if a parser is used which doesn't have support for this.

File: lib/rules/unit-no-unknown/index.js (hunk 2 above)
Comment (romainmenke) [reply lib/rules/unit-no-unknown/index.js]: yes! thanks 🙇

**Review by ybiquitous — APPROVED** [review]
Thank you. LGTM 👍🏼

**Comment by ScorpAL** [comment]
HI Guys.
Something wrong after update from 15.6.2 to 15.6.3 

```css
.test {
    margin-bottom: 20zrr;
}
```
With **stylelint 15.6.2**
CLI and VSCode reported:
`Unexpected unknown unit "zrr"  unit-no-unknown`

With **stylelint 15.6.3**
CLI:
no errors reported
VSCode:
`Unexpected unknown unit "zrr" (unit-no-unknown)Stylelint[unit-no-unknown](https://stylelint.io/user-guide/rules/unit-no-unknown)`

**Pull request merged**